src/pool/pool_disjoint.c

```diff
@@ -0,0 +1,1171 @@
+                                       &pool_size, &new_pool_size)) {
+                if (chunked_bucket) {
+                    ++bucket->chunked_slabs_in_pool;
+                }
+
+                bucket_update_stats(bucket, -1, 1);
+                *to_pool = true;
+                return true;
+            }
+        }
+    }
+
+    bucket_update_stats(bucket, -1, 0);
+    *to_pool = false;
+    return false;
+}
+
+utils_rwlock_t *bucket_get_known_slabs_map_lock(bucket_t *bucket) {
+    return &bucket->pool->known_slabs_map_lock;
+}
+
+static size_t size_to_idx(disjoint_pool_t *pool, size_t size) {
+    assert(size <= CutOff && "Unexpected size");
+    assert(size > 0 && "Unexpected size");
+
+    size_t min_bucket_size = (size_t)1 << pool->min_bucket_size_exp;
+    if (size < min_bucket_size) {
+        return 0;
+    }
+
+    // get the position of the leftmost set bit
+    size_t position = getLeftmostSetBitPos(size);
+
+    bool is_power_of_2 = 0 == (size & (size - 1));
+    bool larger_than_halfway_between_powers_of_2 =
+        !is_power_of_2 &&
+        (bool)((size - 1) & ((uint64_t)(1) << (position - 1)));
+    size_t index = (position - pool->min_bucket_size_exp) * 2 +
+                   (int)(!is_power_of_2) +
+                   (int)larger_than_halfway_between_powers_of_2;
+
+    return index;
+}
+
+umf_disjoint_pool_shared_limits_t *
+disjoint_pool_get_limits(disjoint_pool_t *pool) {
+    if (pool->params.shared_limits) {
+        return pool->params.shared_limits;
+    } else {
+        return pool->default_shared_limits;
+    }
+}
+
+bucket_t *disjoint_pool_find_bucket(disjoint_pool_t *pool, size_t size) {
+    size_t calculated_idx = size_to_idx(pool, size);
+
+#ifndef NDEBUG
+    // debug check
+    bucket_t *bucket = pool->buckets[calculated_idx];
+    assert(bucket->size >= size);
+    (void)bucket;
+
+    if (calculated_idx > 0) {
+        bucket_t *bucket_prev = pool->buckets[calculated_idx - 1];
+        assert(bucket_prev->size < size);
+        (void)bucket_prev;
+    }
+#endif // NDEBUG
+
+    return pool->buckets[calculated_idx];
+}
+
+void bucket_print_stats(bucket_t *bucket, bool *title_printed,
+                        const char *label) {
+    if (bucket->alloc_count) {
+        if (!*title_printed) {
+            LOG_DEBUG("\"%s\" pool memory statistics", label);
+            LOG_DEBUG("%14s %12s %12s %18s %20s %21s", "Bucket Size", "Allocs",
+                      "Frees", "Allocs from Pool", "Peak Slabs in Use",
+                      "Peak Slabs in Pool");
+            *title_printed = true;
+        }
+        LOG_DEBUG("%14zu %12zu %12zu %18zu %20zu %21zu", bucket->size,
+                  bucket->alloc_count, bucket->free_count,
+                  bucket->alloc_pool_count, bucket->max_slabs_in_use,
+                  bucket->max_slabs_in_pool);
+    }
+}
+
+void disjoint_pool_print_stats(disjoint_pool_t *pool, bool *title_printed,
```

Comment:
Instead of bool *title_printed, create function print_title and just print once, without extra pointer argument passed to all functions.